## Introduction

This repository contains the code associated with the first chapter of my PhD at La Trobe University: A project to map patterns of mycorrhizal diversity and endemism in Australian forests.

The project is comprised of seven main tasks:
1. Bioinformatics analysis of fungal (ITS) metabarcoding data from the [Australian Microbiome Initiative (AMI)](https://www.australianmicrobiome.com/)
2. Bioinformatics analysis of bacterial (16S) metabarcoding data from the AMI.
3. Clean and harmonise plant species names from the [Biomass Plot Library](https://portal.tern.org.au/metadata/23218#:~:text=The%20Biomass%20Plot%20Library%20is,private%20companies%20and%20other%20agencies.), [Harmonised Australian Vegetation Plot(HAVPlot)](https://data.csiro.au/collection/csiro:54461?_st=browse&_str=6&_si=2&browseType=kw&browseValue=vegetation), [FungalRoot](https://nph.onlinelibrary.wiley.com/doi/full/10.1111/nph.18207), and [NodDB](https://onlinelibrary.wiley.com/doi/10.1111/jvs.12627) databases.
4. Predict poprtions of root traits (arbuscular mycorrhizal, ectomycorrhizal and nitrogen-fixing) across Australian forests based on the Biomass Plot Library dataset.
5. Extract for environmental and vegetation data for AMI and HAVPlot sites to use in subsequent statistical analyses.
6. Estimate drivers of root trait diversity for both microbes and plants.
7. Determine root trait vulnerability to global change drivers for both microbes and plants.
8. Map patterns of root trait endemism for both microbes and plants.

## 

